[FIRRTL] Improve canonicalization involving `invalidvalue`

fabianschuiki opened the pull request:
This extends the FIRRTL fold and canonicalization patterns to treat `invalidvalue` as a zero if it is an input to a primary operation. This is inline with what the Scala implementation of the FIRRTL compiler does.

Doing this removes quite a few of the special cases where certain ops are aware of `invalidvalue`s and will fold as if it was a zero. The addition of the `getConstant`, `isConstantZero`, and `canonicalizePrimOp` helpers allow for most folders and canonicalizers to not having to bother with `invalidvalue`s, as these helpers implicitly map `invalidvalue`s to zero constants.

This also subsumes the work done in #2278 and #2198, which selectively added handling for `invalidvalue`s to the `PadPrimOp`.

The extension of the folders and canonicalizers now makes all the cases in `invalid-reg-fail.fir` pass, albeit some only after accepting that the MLIR implementation folds *additional* cases where the Scala one might just leave an operation untouched. These are marked with a `<-- fixed; upstream to Scala FIRRTL impl?` and should probably be upstreamed as additional canonicalizations on the Scala side as well.

---
*Original PR:*

This draft PR is an effort to make `invalidvalue` behave as zero during canonicalization, while still maintaining the `firrtl.invalidvalue` operations in the IR, which we use in a canonicalization of `MuxPrimOp` to properly do `WhenOp` expansion.

With a few additional helpers, we can make all the folds and canonicalizations work with `invalidvalue` as zero. However as a result some of the integration tests in `invalid-reg-pass.fir` start to fail, which I think indicates that `firtool` now folds things more aggressively than the Scala implementation.

I don't know if it's worthwhile pursuing this effort, or whether we should rather try to match the Scala impl exactly (which would make logical equivalence pass, but is pretty tricky because we have to be bug-for-bug compatible).

seldridge reviewed (commented):
My gut instinct is that folding _more_ is fine, so long as the fold involves an invalid interpreted as zero. Any error of this form is likely a missing fold in the SFC. (The first failure in `invalid-reg-pass.fir` is related to "convert to signed/cvt" not folding away. This just seems to be missing as SFC doesn't touch `cvt(UInt<1>(0))`.) This type of deviation shouldn't change functional correctness except for pre-reset/cycle zero behavior.

Any fold involving invalid _propagation_ (`foo(_, invalid) -> invalid`) needs to be handled extremely carefully as the SFC never does this. This type of interpretation can introduce functional differences in the output.

fabianschuiki commented:
That makes a lot of sense @seldridge. That would mean some of the test cases in `invalid-reg-pass.fir` would have to be adapted for MFC being able to fold more stuff. We could translate these into issues on the FIRRTL repo as we adjust each case, to ensure we match SFC and MFC without having to pass up on optimization opportunities in MFC.

lattner commented on lib/Dialect/FIRRTL/FIRRTLFolds.cpp:
```diff
@@ -103,6 +104,12 @@ static Optional<APSInt> getExtendedConstant(Value operand, Attribute constant,
 
+  // InvalidValue inputs simply read as zero.
+  if (auto result = constant.dyn_cast_or_null<InvalidValueAttr>()) {
+    llvm::errs() << "Treating " << operand << " as zero value\n";
```
Plz remove logging before committing obviously :-)

lattner commented on lib/Dialect/FIRRTL/FIRRTLFolds.cpp:
```diff
@@ -103,6 +104,12 @@ static Optional<APSInt> getExtendedConstant(Value operand, Attribute constant,
+  // InvalidValue inputs simply read as zero.
+  if (auto result = constant.dyn_cast_or_null<InvalidValueAttr>()) {
+    llvm::errs() << "Treating " << operand << " as zero value\n";
+    return APSInt(destWidth, result.getType().isa<UIntType>());
```
Do invalid's work with reset/clock/arrays and other types?  I am pretty sure it works with aggregates, so you might need to check this (and just `return {}`)

lattner reviewed (commented):
This approach looks good to me!

lattner commented:
I agree with @seldridge here that more folding is probably ok.  That said, I don't think it is important to invest effort in being super aggressive either.

Do we have a good list of things that *really should* be preserved with invalid?  Documenting that will help put guard rails on this and set expectations.  I think the model is pretty simple:

1) Never fold invalid values to zero through connects.
2) When lowering looks at them.
3) They are otherwise treated as zeros.

Does this sound right to you?

seldridge commented:
> Do we have a good list of things that really should be preserved with invalid?

Not exactly other than:

1. A `when` block with a default initialization of invalid should lower to "validif" which we represent as `mux(cond, true, invalid)`.  This has to be preserved and should be preserved by following your "Never fold invalid values to zero through connects".  After expand whens, an invalid value connection can be converted to a connection to zero value.
2. Anything that doesn't violate (1) should be treated as constant zero.

This roughly aligns with what you're saying.

A byproduct of this is that the FIRRTL to HW conversion should see no invalid values if optimizations are on.

lattner commented:
sounds good to me

fabianschuiki commented:
I'll do another pass over this with the findings from the discussion here and elsewhere regarding invalid propagation.

fabianschuiki commented:
@lattner, @seldridge, I've reworked this and extended it to cover all the prim ops. I've also added a bunch of tests for the newly-covered cases, and was able to get rid of @seldridge's XFAIL tests entirely (with a caveat that MFC is now more aggressive in some cases than SFC).

lattner commented:
This looks structurally fine to me, but i'd prefer @seldridge to do a detailed walkthrough

seldridge commented on lib/Dialect/FIRRTL/FIRRTLFolds.cpp:
```diff
@@ -107,6 +108,10 @@ static Optional<APSInt> getExtendedConstant(Value operand, Attribute constant,
   if (destWidth < 0)
     return {};
 
+  // InvalidValue inputs simply read as zero.
+  if (auto result = constant.dyn_cast_or_null<InvalidValueAttr>())
+    return APSInt(destWidth, operand.getType().cast<IntType>().isUnsigned());
+
```
Can you update the the comment on this function to reflect this change in behavior?

fabianschuiki replied:
Done :+1:

seldridge commented on lib/Dialect/FIRRTL/FIRRTLFolds.cpp:
```diff
@@ -118,6 +123,33 @@ static Optional<APSInt> getExtendedConstant(Value operand, Attribute constant,
   return {};
 }
 
+/// Determine the value of a constant operand for the sake of constant folding.
+/// This will map `invalidvalue` to a zero value of the corresopnding type,
+/// which aligns with how the Scala FIRRTL compiler handles invalids.
```
```suggestion
/// Determine the value of a constant operand for the sake of constant folding.
/// This will map `invalidvalue` to a zero value of the corresopnding type,
/// which aligns with how the Scala FIRRTL compiler handles invalids in most
/// cases. For a full discussion of this see the FIRRTL Rationale document.
```

fabianschuiki replied:
Awesome thanks, applied this manually in the rebase.

seldridge commented on lib/Dialect/FIRRTL/FIRRTLFolds.cpp:
```diff
@@ -118,6 +123,33 @@ static Optional<APSInt> getExtendedConstant(Value operand, Attribute constant,
+  if (!operand)
+    return {};
+  if (auto attr = operand.dyn_cast<InvalidValueAttr>()) {
+    if (auto type = attr.getType().dyn_cast<IntType>())
+      return APSInt(type.getWidth().getValueOr(1), type.isUnsigned());
```
Can `getWidth().getValue()` fail here? I assume this is catching the case of this running on an uninferred width?

fabianschuiki replied:
Yeah exactly. My thinking was that there's this `getAPSInt()` call below if we're dealing with an `IntegerAttr`, which would have a width assigned even in the case that the FIRRTLType itself had an uninferred width (just some random width that is sufficient to hold the constant). So in analogy I just made the invalidvalue fold to a zero of at least one bit.

seldridge commented on lib/Dialect/FIRRTL/FIRRTLFolds.cpp:
```diff
@@ -118,6 +123,33 @@ static Optional<APSInt> getExtendedConstant(Value operand, Attribute constant,
+    if (auto type = attr.getType().dyn_cast<IntType>())
+      return APSInt(type.getWidth().getValueOr(1), type.isUnsigned());
+    if (attr.getType().isa<ClockType, ResetType, AsyncResetType>())
+      return APSInt(1, true);
```
`isUnsigned=true` is the default here, so it's probably better to drop the explicit `true`:
```suggestion
      return APSInt(1);
```

Alternatively, `APSInt(1, /*isUnsigned=*/ true)`

fabianschuiki replied:
:+1:

seldridge commented on lib/Dialect/FIRRTL/FIRRTLFolds.cpp:
```diff
@@ -118,6 +123,33 @@ static Optional<APSInt> getExtendedConstant(Value operand, Attribute constant,
+  if (auto attr = operand.dyn_cast<IntegerAttr>())
+    return attr.getAPSInt();
+  if (auto attr = operand.dyn_cast<BoolAttr>())
+    return APSInt(APInt(1, attr.getValue()), true);
```
```suggestion
    return APSInt(APInt(1, attr.getValue()));
```

fabianschuiki replied:
:+1:

seldridge commented on lib/Dialect/FIRRTL/FIRRTLFolds.cpp:
```diff
@@ -192,6 +224,63 @@ constFoldFIRRTLBinaryOp(Operation *op, ArrayRef<Attribute> operands,
+static LogicalResult canonicalizePrimOp(
+    Operation *op, PatternRewriter &rewriter,
+    const function_ref<OpFoldResult(ArrayRef<Attribute>)> &canonicalize) {
+  // Can only operate on FIRRTL primary operations.
```
Nit: I think this should be "primitive" and not "primary"?

fabianschuiki replied:
Yeah totally, fixed.

seldridge commented on lib/Dialect/FIRRTL/FIRRTLFolds.cpp:
```diff
@@ -192,6 +224,63 @@ constFoldFIRRTLBinaryOp(Operation *op, ArrayRef<Attribute> operands,
+  for (auto operand : op->getOperands()) {
+    Attribute attr;
+    if (auto *defOp = operand.getDefiningOp())
+      TypeSwitch<Operation *>(defOp)
+          .Case<ConstantOp, SpecialConstantOp, InvalidValueOp>(
+              [&](auto op) { attr = op.fold({}).template get<Attribute>(); });
```
MSVC may not like the template inside a lambda. You may want to do a special run with the MSVC action using this branch to make sure it compiles. If this is a problem, I will usually work around it by removing the `auto`. In this case it's necessary as there's no parent type that can be used here. Worst case, this could use three cases with explicit types instead of `auto`.

fabianschuiki replied:
Good idea, will do a run!

seldridge commented on lib/Dialect/FIRRTL/FIRRTLFolds.cpp:
```diff
@@ -237,30 +326,51 @@ OpFoldResult AddPrimOp::fold(ArrayRef<Attribute> operands) {
-                                            MLIRContext *context) {
-  results.insert<patterns::AddWithInvalidOp>(context);
+LogicalResult AddPrimOp::canonicalize(AddPrimOp op, PatternRewriter &rewriter) {
+  return canonicalizePrimOp(op, rewriter,
+                            [&](ArrayRef<Attribute> operands) -> OpFoldResult {
+                              // add(x, 0) -> x
+                              // add(0, x) -> x
+                              if (isConstantZero(operands[1]))
+                                return op.getOperand(0);
+                              if (isConstantZero(operands[0]))
+                                return op.getOperand(1);
+                              return {};
+                            });
```
Nit: it may be more readable to move the comment onto the `if` statement it correlates with as opposed to putting them all at the top. Ditto for below.

fabianschuiki replied:
Yeah good thinking.

seldridge commented on lib/Dialect/FIRRTL/FIRRTLFolds.cpp:
This references division? Do we actually do this fold for remainder, too?

fabianschuiki replied:
Yeah, there's a `div(x, x) -> 1` and a `rem(x, x) -> 0)` fold which cater to the same idea. This is one of the things that aren't there in SFC, and it would be an easy fix to just disable the folds here to make MFC match up with SFC.

seldridge commented on lib/Dialect/FIRRTL/FIRRTLFolds.cpp:
```diff
@@ -383,20 +491,7 @@ OpFoldResult AndPrimOp::fold(ArrayRef<Attribute> operands) {
-    return lhs();
-
-  if (auto rhsCst = operands[1].dyn_cast_or_null<IntegerAttr>()) {
+  if (auto rhsCst = getConstant(operands[1])) {
```
This now only has to check the RHS because `or(invalid, x)` is properly canonicalized to `or(x, invalid)` because invalid is zero-like?

fabianschuiki replied:
Yeah exactly. The `OrPrimOp` is marked as commutative, and the `InvalidValueOp` folds to a `InvalidValueAttr` (which I guess MLIR interprets as "the constant"), so a builtin pattern will swap the operands on that `OrPrimOp`.

seldridge reviewed (approved):
This looks awesome. I love how everything unifies around `isConstant`.

I will open issues about all of this on the SFC.  The only ones that I'm slightly nervous about is places where we use `specialconstant` for clock and reset literals.  The SFC doesn't have these, so there will always be `asClock` or `asAsyncReset` casts that it can't remove. (This is a place where these ops would block register removal.)  This isn't enough to block these until we have a problem, thouhg.

seldridge commented:
Tracking issue on the SFC about the observed differences: https://github.com/chipsalliance/firrtl/issues/2432

fabianschuiki commented on include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td:
```diff
@@ -142,33 +135,6 @@ def GetWidthAsIntAttr : NativeCodeCall<
-    (KnownWidth $x), (IsInvalid $y)
-  ]>;
-
-// pad(invalid, width) -> zero<width>
-def PadInvalid : Pat<
-  (PadPrimOp:$result (InvalidValueOp), $_),
-  (GetZeroConstant $result), []>;
```
Yeah I was happy that we no longer needed some of the explicit patterns after this change. :-D

youngar commented:
There is a bug in HEAD compiling regression/test3 that this might fix, `GetZeroConstant` seems to always return a unsigned constant. Smaller test case here:
```mlir
 firrtl.circuit "Y" {
firrtl.module @Y(in %i: !firrtl.sint<1>, out %o: !firrtl.sint<15>) {
  %invalid_si1 = firrtl.invalidvalue : !firrtl.sint<1>
  %0 = firrtl.pad %invalid_si1, 15 : (!firrtl.sint<1>) -> !firrtl.sint<15>
  firrtl.connect %o, %0 : !firrtl.sint<15>, !firrtl.sint<15>
}
}
```

fabianschuiki commented:
Thanks for checking @youngar. I've added your test case as a regression test to `canonicalization.mlir` :+1:.
